File: design_patterns/builder/Builder.java

class Builder {

    private int a = 0;
    private int b = 0;

    public Builder addA(int a) {
        this.a = a;
        return this;
    }

    public Builder addB(int b) {
        this.b = b;
        return this;
    }

    @Override
    public String toString() {
        return "Builder { a: " + this.a + ", b: " + this.b + " }";
    }

    public static void main(String[] args) {

        Builder builder = new Builder().addA(10).addB(20);

        System.out.println(builder);
    }
}

File: design_patterns/builder/builder.cpp
#include <iostream>
#include <ostream>

class Thing {
private:
    int a = 0;
    int b = 0;

public:
    Thing& addA(int a) {
        this->a = a;
        return *this;
    }

    Thing& addB(int b) {
        this->b = b;
        return *this;
    }

    friend std::ostream& operator<<(std::ostream& os, Thing const& thing) {
        os << "Thing { a: " << thing.a << ", b: " << thing.b << " }";
        return os;
    }

};


int main() {

    Thing thing;
    thing.addA(10).addB(20);

    std::cout << thing << "\n";

    return 0;
}
